[Security Solution] Fix flaky test: x-pack/test/detection_engine_api_integration/security_and_spaces/update_prebuilt_rules_package/update_prebuilt_rules_package·ts - update_prebuilt_rules_package should allow user to install prebuilt rules from scratch, then install new rules and upgrade existing rules from the new package

jpdjere opened the pull request:
Fixes: https://github.com/elastic/kibana/issues/162658

## Summary

- Fixes flaky test: `x-pack/test/detection_engine_api_integration/security_and_spaces/update_prebuilt_rules_package/update_prebuilt_rules_package·ts`
- Test title: `update_prebuilt_rules_package should allow user to install prebuilt rules from scratch, then install new rules and upgrade existing rules from the new package`

## Passing flaky test runner

300 runs: https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds/2799

## Root cause and what this PR does

- On initial investigation, the flaky test runner was executed for this test with 300 iterations and all of them succeeded. This gives us great confidence that the test is not actually flaky.
- Further investigation showed that @kibanamachine reported this tests as failing when running the CI for two backport PRs to `8.9`:
    - https://buildkite.com/elastic/kibana-on-merge/builds/33282#0189987d-3a80-49c2-8332-3105ec3c2109
    - https://buildkite.com/elastic/kibana-on-merge/builds/33444#0189b1fa-4bc4-4422-9ce9-5c9a24f11ad5
- These flakiness was caused **by a race condition** between the writing of rules into indeces, and them being made available for reading. This is a known issue, already and solved for other integration test, by manually refreshing ES's indeces. 
- In order to reduce the probability of flakiness in a similar scenario, this PR adds code to refresh the indices after each  rule installation or upgrade during the test.

## Refactor

- Moves the refreshing of the indexes within the utility function that write to indexes:
    - `installPrebuiltRules`
    - `upgradePrebuiltRules`
    - `installPrebuiltRulesAndTimelines` (legacy, but still used)
    - `installPrebuiltRulesFleetPackage`
- Creates 2 new utils:
    - `installPrebuiltRulesPackageByVersion`, which installs `security_detection_engine` package via Fleet API, with its version passed in as param
    - `getInstalledRules`, reusable function to fetch all installed rules
    -

jpdjere commented:
https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds/2795

jpdjere commented:
Second attempt - 300 runs: https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds/2799

elasticmachine commented:
Pinging @elastic/security-detections-response (Team:Detections and Resp)

elasticmachine commented:
Pinging @elastic/security-solution (Team: SecuritySolution)

maximpn commented on x-pack/test/detection_engine_api_integration/security_and_spaces/update_prebuilt_rules_package/update_prebuilt_rules_package.ts:
```diff
@@ -133,6 +134,8 @@ export default ({ getService }: FtrProviderContext): void => {
       // Verify that the _perform endpoint returns the same number of installed rules as the status endpoint
       // and the _review endpoint
       const installPrebuiltRulesResponse = await installPrebuiltRules(supertest);
+      await es.indices.refresh({ index: ALL_SAVED_OBJECT_INDICES });
```
It's quite easy to forget why `es.indices.refresh()` is needed and what part of test it's related to. It'd be nicer to have it right before an HTTP call depending on it so anybody reusing it doesn't have to keep in mind `es.indices.refresh()` has to be called before. For example `getPrebuiltRulesStatus()` should contain `es.indices.refresh()` inside.

maximpn commented on x-pack/test/detection_engine_api_integration/security_and_spaces/update_prebuilt_rules_package/update_prebuilt_rules_package.ts:
```diff
@@ -133,6 +134,8 @@ export default ({ getService }: FtrProviderContext): void => {
       // and the _review endpoint
       const installPrebuiltRulesResponse = await installPrebuiltRules(supertest);
+      await es.indices.refresh({ index: ALL_SAVED_OBJECT_INDICES });
+
```
Installed rules fetching chunk looks to be repeated. It could be a reusable function with `es.indices.refresh()` inside

```
const { body: rulesResponse } = await supertest
        .get(`${DETECTION_ENGINE_RULES_URL_FIND}?per_page=10000`)
        .set('kbn-xsrf', 'true')
        .send()
        .expect(200);
```

maximpn reviewed (commented):
@jpdjere thank you for fixing this flaky test 👍 

Could you describe the root cause in the ticket's description? I'd be good to isolate `es.indices.refresh()` to the utility functions where it's needed and have a clear comment describing why it's needed there.

Have you checked also all the above requests have `?refresh=wait_for` so the changed data is available? I'm a bit afraid we don't make it clear here in tests so any further readers will be confused.

jpdjere commented:
@maximpn Refactored the utils according to your feedback.

In the case of utility functions that write to an index, I moved the refresh method to within the util and added a comment of why there. Notice that this is done **after writing to the index** for each case, to leave the test in a state in which it is ready and updated for any subsequent read operation and assertion on those results.

Also moved the "find rules" to its own utility `getInstalledRules`; this one doesn't need an index refresh since it's a read operation.

maximpn commented on x-pack/test/detection_engine_api_integration/utils/prebuilt_rules/get_installed_rules.ts:
```diff
@@ -0,0 +1,30 @@
+ * @returns Fleet install package response
+ */
+
+export const getInstalledRules = async (
```
nit: IMHO `fetchInstalledRules` would be the best name for the functionality under the hood as it clearly describes that it fetches the data presumable from an endpoint. While `get` usually used for sync functions extracting some data like `Map.get`. At the same time `getInstalledRules` conforms with the other functions in the folder like `getPrebuiltRulesStatus`. Ideally `getPrebuiltRulesStatus` should be renamed to `fetchPrebuiltRulesStatus` as well as the others but it definitely out of scope of this PR.

jpdjere replied:
Makes sense. I'll address this in any of the following PRs I'm working on. Thanks for all the feedback!

maximpn reviewed (approved):
@jpdjere thank you for addressing my comments 🙏
